table/block.cc

```diff
@@ -524,12 +590,40 @@ Block::Block(BlockContents&& contents, SequenceNumber _global_seqno,
     // Should only decode restart points for uncompressed blocks
     if (compression_type() == kNoCompression) {
       num_restarts_ = NumRestarts();
-      restart_offset_ =
-          static_cast<uint32_t>(size_) - (1 + num_restarts_) * sizeof(uint32_t);
-      if (restart_offset_ > size_ - sizeof(uint32_t)) {
-        // The size is too small for NumRestarts() and therefore
-        // restart_offset_ wrapped around.
-        size_ = 0;
+      switch (IndexType()) {
+        case BlockBasedTableOptions::kDataBlockBinarySearch:
+          restart_offset_ = static_cast<uint32_t>(size_) -
+                            (1 + num_restarts_) * sizeof(uint32_t);
+          if (restart_offset_ > size_ - sizeof(uint32_t)) {
+            // The size is too small for NumRestarts() and therefore
+            // restart_offset_ wrapped around.
+            size_ = 0;
+          }
+          break;
+        case BlockBasedTableOptions::kDataBlockHashSearch:
+          if (size_ < sizeof(uint32_t) /* NUM_RESTARTS*/ +
+                          sizeof(uint16_t) * 2 /* NUM_BUCK and MAP_START */) {
+            size_ = 0;
+            break;
+          }
+
+          data_block_hash_index_.reset(new DataBlockHashIndex(
+              Slice(contents.data.data(), /* chop off NUM_RESTARTS */
+                    contents.data.size() - sizeof(uint32_t))));
```

Comment:
We should think about whether we can get rid of this malloc. Is there a way to limit things we need to load for hash index to small size and just keep it in class Block?